### PR TITLE
Aspects: Create exception for aspects lookup

### PR DESCRIPTION
**Pull request by adhikasp**
Closes https://github.com/coala/coala/issues/4381

### CODE REVIEW

File: coalib/bearlib/aspects/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+class aspectTypeError(TypeError):
```
Comment (Monal5031) [inline coalib/bearlib/aspects/exceptions.py]: I think the name should be `AspectTypeError` as to follow python pep8 naming guideines.

File: coalib/bearlib/aspects/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    """
+
+    def __init__(self, item):
+        self.item = item
+        message = ('{} is not an aspectclass or an instance of an '
+                   'aspectclass'.format(repr(self.item)))
+        super().__init__(message)
+
+
+class aspectLookupError(LookupError):
```
Comment (Monal5031) [inline coalib/bearlib/aspects/exceptions.py]: Also this should be `AspectLookupError`.

File: coalib/bearlib/aspects/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    def __init__(self, aspectname, message=None):
+        self.aspectname = aspectname
+        if message is None:
+            message = ('error when trying to search aspect named {}'
+                       .format(repr(aspectname)))
+        super().__init__(message)
+
+
+class aspectNotFoundError(aspectLookupError):
```
Comment (Monal5031) [inline coalib/bearlib/aspects/exceptions.py]: Also this as `AspectNotFoundError`.

File: coalib/bearlib/aspects/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    """
+
+    def __init__(self, aspectname):
+        message = ('no aspect named {}'.format(repr(aspectname)))
+        super().__init__(aspectname, message)
+
+
+class MultipleaspectFoundError(aspectLookupError):
+    """
+    Multiple aspect is found.
```
Comment (Monal5031) [inline coalib/bearlib/aspects/exceptions.py]: typo here it should be `aspects`.

File: coalib/bearlib/aspects/exceptions.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    """
+    No aspect found.
+    """
+
+    def __init__(self, aspectname):
+        message = ('no aspect named {}'.format(repr(aspectname)))
+        super().__init__(aspectname, message)
+
+
+class MultipleaspectFoundError(aspectLookupError):
```
Comment (Monal5031) [inline coalib/bearlib/aspects/exceptions.py]: Also this should be `MultipleAspectFoundError`.

File: coalib/bearlib/aspects/__init__.py [hunk 6]
```diff
@@ -6,7 +6,9 @@
 from types import ModuleType
 
 from .base import aspectbase
-from .meta import aspectclass, aspectTypeError
+from .exceptions import (aspectTypeError, aspectNotFoundError,
```
Comment (Monal5031) [inline coalib/bearlib/aspects/__init__.py]: each module on different like would look good here.

File: coalib/bearlib/aspects/exceptions.py (hunk 5 above)
Comment (jayvdb) [inline coalib/bearlib/aspects/exceptions.py]: This is a great example of how ridiculous the 'lowercase a in aspect' thing is.

**Comment by adhikasp** [comment]
Update: partly fix https://github.com/coala/coala/issues/4402

File: tests/bearlib/aspects/CollectionsTest.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import pytest
 
-from coalib.bearlib.aspects import aspectTypeError
+from coalib.bearlib.aspects import ApectTypeError
```
Comment (jayvdb) [inline tests/bearlib/aspects/CollectionsTest.py]: ... OR ... use `from coalib.bearlib.aspects import \n    (ApectTypeError as aspectTypeError)` so you do not need to modify the `pytest.raises`

File: coalib/bearlib/aspects/meta.py [hunk 8]
```diff
@@ -76,19 +77,6 @@ def __repr__(cls):
         return '<%s %s>' % (type(cls).__name__, repr(cls.__qualname__))
 
 
-class aspectTypeError(TypeError):
```
Comment (jayvdb) [inline coalib/bearlib/aspects/meta.py]: replace with a backwards compatible

`aspectTypeError = AspectTypeError`

... OR ... see below

File: tests/bearlib/aspects/ClassTest.py [hunk 9]
```diff
@@ -1,4 +1,4 @@
-from coalib.bearlib.aspects import Root, aspectTypeError
+from coalib.bearlib.aspects import Root, ApectTypeError
```
Comment (jayvdb) [inline tests/bearlib/aspects/ClassTest.py]: ... OR ... use `from coalib.bearlib.aspects import \n    (ApectTypeError as aspectTypeError)` so you do not need to modify the `pytest.raises` here and on the next file.

Reason is they will be updated anyway in https://github.com/coala/coala/pull/3894
(And after this is merged, review that PR and ensure that they update the name there.)

File: coalib/bearlib/aspects/exceptions.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    def __init__(self, item):
+        self.item = item
+        message = ('{} is not an aspectclass or an instance of an '
+                   'aspectclass'.format(repr(self.item)))
+        super().__init__(message)
+
+
+class AspectLookupError(LookupError):
+    """
+    Error that raised when trying to search aspect.
```
Comment (jayvdb) [inline coalib/bearlib/aspects/exceptions.py]: Error that raised -> Error that is raised ..

or -> Error raised ..

But "No aspect found." is more in the same style as your other docstrings.

Your choice ;-)

File: tests/bearlib/aspects/ClassTest.py [hunk 11]
```diff
@@ -1,4 +1,4 @@
-from coalib.bearlib.aspects import Root, aspectTypeError
+from coalib.bearlib.aspects import (Root, AspectTypeError as aspectTypeError)
```
Comment (jayvdb) [inline tests/bearlib/aspects/ClassTest.py]: round brackets not needed if `\n` not added ? ?

File: coalib/bearlib/aspects/exceptions.py (hunk 10 above)
Comment (adhikasp) [inline coalib/bearlib/aspects/exceptions.py]: I should really consider a grammar lesson :(

Comment (jayvdb) [inline coalib/bearlib/aspects/exceptions.py]: Not fixed?

File: tests/bearlib/aspects/ExceptionsTest.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+from coalib.bearlib.aspects.exceptions import AspectLookupError
+
+
+class ExceptionTest:
+
+    def test_AspectLookupError(self):
+        with pytest.raises(AspectLookupError) as exc:
+            raise AspectLookupError('NOASPECT')
+        assert str(exc.value) == \
```
Comment (jayvdb) [inline tests/bearlib/aspects/ExceptionsTest.py]: please do not use line continuation ( `\` ) , as it is a parsing nightmare.  Using round brackets.

Comment (jayvdb) [inline tests/bearlib/aspects/ExceptionsTest.py]: also use `self.assertEqual`

File: tests/bearlib/aspects/ExceptionsTest.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import pytest
+
+from coalib.bearlib.aspects.exceptions import AspectLookupError
+
+
+class ExceptionTest:
+
+    def test_AspectLookupError(self):
+        with pytest.raises(AspectLookupError) as exc:
```
Comment (jayvdb) [inline tests/bearlib/aspects/ExceptionsTest.py]: do not use pytest specific functions in the unit tests.  We have an issue to remove them.

**Comment by jayvdb** [comment]
ack d17af96 4b68cf3

File: tests/bearlib/aspects/ExceptionsTest.py [hunk 14]
```diff
@@ -0,0 +1,11 @@
+import unittest
+
+from coalib.bearlib.aspects.exceptions import AspectLookupError
+
+
+class ExceptionTest(unittest.TestCase):
+
+    def test_AspectLookupError(self):
+        with self.assertRaisesRegex(
+                AspectLookupError, 'Error when trying to search aspect'):
```
Comment (jayvdb) [inline tests/bearlib/aspects/ExceptionsTest.py]: This should be a more precise regex - i.e. verifying an exact string `^...$`.

**Comment by jayvdb** [comment]
reack b219eb6 0ef6e01

**Comment by jayvdb** [comment]
@rultor merge

**Pull request merged**